core/src/main/java/com/turn/ttorrent/client/Piece.java

```diff
@@ -172,6 +196,37 @@ public synchronized boolean validate() throws IOException {
 	}
 
 	/**
+	 * Internal piece data read function without memory allocation.
+	 *
+	 * <p>
+	 * This function will read the piece data without checking if the piece has
+	 * been validated. It is simply meant at factoring-in the common read code
+	 * from the validate and read functions.
+	 * </p>
+	 *
+	 * @param offset Offset inside this piece where to start reading.
+	 * @param buffer A byte buffer to read the piece data into.
+	 * @throws IllegalArgumentException If <em>offset + length</em> goes over
+	 * the piece boundary.
+	 * @throws IOException If the read can't be completed (I/O error, or EOF
+	 * reached, which can happen if the piece is not complete).
+	 */
+	private void _read(long offset, ByteBuffer buffer) throws IOException {
```

Comment:
Can't you call this method from the other `_read()` method with the newly allocated `ByteBuffer`?